### ISSUE TITLE
Understanding evolve results

### ISSUE THREAD

**Issue by nahidalam** [ultralytics/yolov5#11582]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

I am running sample `evolve` experiments. Attached graph results I am trying to understand. Why does on the validation set, the loss remains zero and at the end of the epoch (50 in this case) shoots up?

<img width="1114" alt="Screenshot 2023-05-23 at 10 49 57 AM" src="https://github.com/ultralytics/yolov5/assets/2341472/d8b44fdb-47ed-4ed5-9454-4cbf7cf7fab2">

Similar things happen for metric/mAP, precision, recall etc.

<img width="1092" alt="Screenshot 2023-05-23 at 11 09 43 AM" src="https://github.com/ultralytics/yolov5/assets/2341472/f6484201-c1eb-42ff-b2b7-c02568c67427">

**Comment by glenn-jocher** [comment]
@nahidalam hello! The observed behavior is quite normal for the evolution process. During the training phase, the weights of the neural network are updated after every batch of images. At the end of the epoch, the average metric values for the whole epoch are computed and used to update the optimizer parameters. 

Thus, the discrepancy between the validation loss in the last batch and at the end of the epoch occurs because during the last batch, the weights are not updated. 

I hope this helps, feel free to ask any further questions!

**Comment by nahidalam** [comment]
Thanks for the quick reply @glenn-jocher 

So what is a good way to decide on which evolve run is the best?

**Comment by glenn-jocher** [comment]
@nahidalam you're welcome!

When deciding on the best evolve run, it is important to look at both the metrics (mAP, precision, recall, etc.) as well as the validation loss. You should choose the run that has the highest mAP and the lowest validation loss.

However, it is important to beware of overfitting, which occurs when a model performs well on the training dataset but poorly on the validation dataset. To avoid overfitting, it is recommended to choose the run that has the best performance on the validation dataset, as this indicates that the model is generalizing well to unseen data.

I hope this helps! Let me know if you have any further questions or concerns.

**Comment by nahidalam** [comment]
Thanks @glenn-jocher, I understand what you just described above from the fundamental perspective. But for `evolve` run, since the validation loss and mAP are both `0` across all the epochs except the last one, that basically means I should take the decision based on the result at the last epoch?

**Comment by glenn-jocher** [comment]
@nahidalam hello!

Yes, you are correct. Since the validation loss and mAP were both zero across all the epochs except the last one, your decision should be based on the results of the last epoch. 

In this case, you should choose the model that has the highest mAP and lowest validation loss in the last epoch, as this indicates that the model has the best performance on the validation dataset and is most likely to generalize well to unseen data.

I hope this helps! Let me know if you have any further questions.
